### ISSUE TITLE
Preserve value order in ContextExtension on serialization

### ISSUE THREAD

**Issue by greenhat** [ScorexFoundation/sigmastate-interpreter#681]
### Problem
ContextExtension values order is not preserved in JSON and binary serializaion. Since ContextExtension serialized values used in transaction id generation it should lead to unstable tx id meaning that the same tx would have different ids after JSON serialization/deserialization.

### Solution
Backward compatibility requirements:
- have to be encoded as a dictionary in JSON (no order preservation);
- have to parse unordered values from bytes;

Assuming the order is lost on JSON roundtrip just switching to order preserving structure to store values is not enough.
Since only the order in binary serialization is critical (included in tx id), I suggest to sort values by keys on binary serialization.

**Comment by aslesarenko** [comment]
I assume deserializer should also enforce ordering requirement, otherwise roundtrip property is not preserved. 
This means the new deserializer should throw exception when order is violated.
This cannot be done in v3.x, so I suggest to change to v4.0 milestone.

**Comment by greenhat** [comment]
@aslesarenko I don't see why binary roundtrip property is important. Ordering is important only in binary serialization (for tx id). In other cases, it used as a key-value map, so the order is not important.

**Comment by aslesarenko** [comment]
@greenhat Well, then I lost the point. 
Could you please clarify the description: 
1) which part of the code this issue relates to
2) what exactly is the problem (give an example please)

**Comment by aslesarenko** [comment]
@greenhat I think roundtrip property is important for consistency of the codebase (since all serializers have this property)
Also soundness of the equality is important.
Unsound equality potentially leads to very subtle bugs.

I'm definitely willing to fix this in v4.x.

**Comment by greenhat** [comment]
@aslesarenko Yes, you're right. I agree that binary roundtrip property is important here for consistency and soundness. I meant that it does not fix the issue - getting values in the different order after a JSON roundtrip.
I assume we cannot fix the order in JSON for compatibility reasons.

**Comment by kushti** [comment]
@aslesarenko @greenhat  I think there's no need to enforce ordering on deserialization, so to change the protocol. Enforcing ordering should be done in JSON and on binary serialization likely. but it is up to a client.

For deserialization, we need to document what happens when two value with the same id come (which one is going to be suppressed), to simplify clients development.

**Comment by greenhat** [comment]
@kushti Enforcing ordering in JSON would mean changing its format from map to array. Which seems like a breaking change. Can we afford it? 
I don't mind leaving binary deserialization as is (no enforcing of the order).

**Comment by kushti** [comment]
@greenhat you mean interaction between client apps (of different version) can be broken? maybe , but not the protocol . and no apps are using many context extension values likely

**Comment by aslesarenko** [comment]
Seems all agree to leave it as is. Closing it then.
